rpc: add locality to dialing

andrewbaptist opened the pull request:
Previously RPC metrics were broken down by individual remote node and class if `server.child_metrics.enabled` was set to true. This set of commits changes the behavior to publish the RPC network metrics by locality instead. 

Epic: CRDB-41138

Release note (ops change): Adds three new network tracking metrics. `rpc.connection.connected` is the number of rRPC TCP level connection established to remote nodes. `rpc.client.bytes.egress` is the number of TCP bytes sent via gRPC on connections we initiated. `rpc.client.bytes.ingress` is the number of TCP bytes received via gRPC on connections we initiated.

Commit:
rpc: add Locality to dialing

Previously we didn't include the locality of the remote node when we
dialed a node. This prevented us from capturing locality aware stats for
the connections.

Epic: CRDB-41138

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126150)
<!-- Reviewable:end -->

Commit:
rpc: allow splitting network stats by locality

Previously we would publish network stats broken down by every remote
node. This would result in a large number of stats for large clusters.
In practice we can aggregate them by remote localities. This reduces
the number of stats with only a minimal loss of visibility into how the
network is being used.

Epic: CRDB-41138

Release note: None

Commit:
rpc: pass Metrics by pointer

The Metrics object stores all the metrics for the connection and was
previously passed by value. This PR passes it by pointer instead to
allow more complex state within the Metrics object in future commits.

Epic: CRDB-41138

Release note: None

Commit:
rpc: add Locality to UnvalidatedDial

Some of the places that call UnvalidatedDial have the locality. By
passing it in when it is known they will more accurately update the
statistics.

Epic: CRDB-41138

Release note: None

Commit:
rpc: add Locality to the ContextOptions

This commit adds the nodes locality information into the ContextOptions.
This allows metrics to consult this to determine if a connection is
from a remote locality.

Epic: CRDB-41138

Release note: None

Commit:
rpc: add metrics for connected count and bytes sent

Previously we didn't track bytes sent and received per node. This commit
adds the metrics for these. Additionally it adds a metric for the
connected count from a TCP perspective as this may be different than the
healthy or unhealty counts.

Epic: CRDB-41138

Release note (ops change): Adds three new network tracking metrics.
`rpc.connection.connected` is the number of rRPC TCP level connections
established to remote nodes. `rpc.client.bytes.egress` is the number of
TCP bytes sent via gRPC on connections we initiated.
`rpc.client.bytes.ingress` is the number of TCP bytes received via gRPC
on connections we initiated.

Commit:
rpc: extract constant for expected count

Extract a constant to make it easier to change the expected count.

Epic: CRDB-41138

Release note: None

Commit:
rpc: more precise peer metrics

Previously the metrics for ConnectionHealthy, ConnectionUnhealthy and
ConnectionInactive were manually set to 0 or 1. This prevented easily
aggregating the peer metrics by something (like locality). This PR
changes the way those three metrics are handled to only increment or
decrement rather than setting to 0/1.

Epic: CRDB-41138

Release note: None

Commit:
rpc: retain metrics in a map

Retain metrics across dropped/re-established connections. If we delete
and unlink the counters if we later recreate it the counts will start at
zero after a reconnect. Instead we track the counters in a map and reuse
them later if the key matches.

Epic: CRDB-41138

Release note: None

Commit:
rpc: make the roundTripLatency metric threadsafe

Previously the metric was not threadsafe and this prevented it from
being shared by multiple connections. The value is only updated on
heartbeat messages, so adding syncronization here should not cause any
performance issues.

Epic: CRDB-41138

Release note: None

andrewbaptist commented on pkg/rpc/metrics.go:
```diff
@@ -249,50 +260,66 @@ type peerMetrics struct {
+			pm.roundTripLatency.Set(0)
+			pm.ConnectionConnected.Update(0)
+
+			pm.ConnectionHealthy.Unlink()
```
Note: Don't unlink when we get to 0.

andrewbaptist commented on pkg/rpc/metrics.go:
```diff
@@ -204,6 +213,8 @@ type peerMetrics struct {
 	// See TestMetricsRelease.
 	release func()
 
+	registeredCount *int
```
We still need it if we want to update the gauges to 0. I'm not positive that it matters if we reset them to 0. So we could potentially remove the registered count completely, but I left it for now.

andrewbaptist commented on pkg/rpc/metrics.go:
```diff
@@ -134,9 +134,9 @@ over this connection.
 )
 
-func makeMetrics() Metrics {
+func makeMetrics() *Metrics {
```
Changed to newMetrics

jeffswenson commented on pkg/rpc/metrics.go:
```diff
@@ -108,33 +112,130 @@ is reset to zero.
+	mu                            struct {
+		syncutil.Mutex
+		// peerMetrics is a map of peerKey to peerMetrics.
+		peerMetrics map[string]peerMetrics
```
If we are no longer changing peerMetrics, do we still need this map of metics and the mutex changes to the moving average latency?

andrewbaptist replied:
The map of metrics helps by preventing the "restart" of metrics when we disconnect and then reconnect. This was what you had mentioned last week and it actually makes the entire flow simpler since we can keep them in a map.

I was able to remove the reference counting and release functions completely by making the metrics more precise (we are always at 0 for the metrics we care about. We don't necessarily need the mutex around the moving average but I expect that now that we have the ability to do arbitrary aggregation of the metrics it is cleaner. I pushed a new set of smaller commits, but I'm going to wait until they pass on CI and do a simple test to make sure it all looks good.

jeffswenson reviewed (approved):
LGTM

Commit:
rpc: skip TestTenantGRPCServices under stress

This test flakes under stress/race conditions due to the use of network
ports. Skipping under stress.

Epic: None

Release note: None

andrewbaptist commented:
TFTR! 

I validated that `TestTenantGRPCServices` was failing under stress/race prior to my commit, so I excluded it. 

bors r=jeffswenson
